modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+{
+	int result = __real_location_request(config);
+
+	if (result == 0) {
+		memfault_gnss_metrics_start_fix_session();
+	}
+	return result;
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(
+	__builtin_types_compatible_p(__typeof__(&location_request),
+				     __typeof__(&__wrap_location_request)) &&
+		__builtin_types_compatible_p(__typeof__(&location_request),
+					     __typeof__(&__real_location_request)),
+	"Error: Wrapped function does not match original location_request function signature");
+
+static void memfault_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_LOCATION || event_data->id == LOCATION_EVT_TIMEOUT ||
+	    event_data->id == LOCATION_EVT_ERROR || event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		memfault_gnss_metrics_stop_fix_session(event_data);
+	}
+	real_location_event_handler(event_data);
+}
+
+/**
+ * @brief Start a GNSS fix session.
+ *
+ */
+static void memfault_gnss_metrics_start_fix_session(void)
+{
+	if (session_in_progress) {
+		return;
+	}
+
+	session_in_progress = true;
+	LOG_DBG("Starting GNSS session");
+	MEMFAULT_METRICS_SESSION_START(ncs_gnss);
+	MEMFAULT_METRIC_ADD(ncs_gnss_fix_request_count, 1);
+}
+
+/**
+ * @brief Stop a GNSS fix session.
+ *
+ * @param event_data pointer to location event data
+ */
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data)
+{
+	if (!session_in_progress) {
+		return;
+	}
+
+	uint32_t session_time_ms;
+	memfault_metrics_heartbeat_timer_read(
+		MEMFAULT_METRICS_KEY(MEMFAULT_METRICS_SESSION_TIMER_NAME(ncs_gnss)),
+		&session_time_ms);
+
+	switch (event_data->id) {
+	case LOCATION_EVT_LOCATION:
+		LOG_DBG("Stopping GNSS session, fix data acquired");
+
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_time_to_fix_ms, session_time_ms);
+
+		uint32_t accuracy_cm = (uint32_t)(event_data->location.accuracy * 100.0f);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_fix_accuracy_cm, accuracy_cm);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_satellites_tracked_count,
+					     event_data->location.details.gnss.satellites_tracked);
+		break;
+	case LOCATION_EVT_TIMEOUT:
+		LOG_DBG("Stopping GNSS session, timeout recorded");
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_search_timeout_ms, session_time_ms);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_satellites_tracked_count,
+					     event_data->location.details.gnss.satellites_tracked);
+		break;
+	case LOCATION_EVT_RESULT_UNKNOWN:
+		// This location result will occur when a timeout occurs, but another method is
+		// attempted and fails (some "external" method via A-GNSS, P-GPS, LTE neighbor
+		// cell and Wi-Fi access point data). Record as a timeout as well but mark with
+		// a different log message for debugging purposes.
+		LOG_DBG("Stopping GNSS session, timeout recorded along with unknown result");
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_gnss_search_timeout_ms, session_time_ms);
```

Comment:
Does the search timeout length vary or is it configured in code? Thinking maybe it's less important to collect if it is configured statically. On the other hand if it is changed frequently or depends on the type of error, etc then it seems like a good idea to keep

Reply:
There are both compile time and run-time options for the timeouts. Applications could set the timeout differently based on whatever criteria when they call location_request() but generally it's very much hard coded, and mainly the Location library default value is used in practice.
Please also notice that there is a timeout for the entire location request, and separate timeouts for individual methods (GNSS, Wi-Fi, cellular).

Reply:
Even though it is generally hard-coded, I think reporting how long we searched for would be useful for later analysis on how that how differences in the timeout might correlate to parameters in the rest of the system (battery life, for example). Furthermore, we need some way to indicate that a session was unsuccessful, and I think the timeout value is more useful than, say, a boolean indicating the session's success. I'm less passionate about method-specific timeouts, though I would not object to also having that passed to the hander somehow through the location_data_details struct.

I am less familiar with the Wi-Fi and cellular methods for getting location. We are definitely honed in on GNSS metrics at the moment. However, in your experience, do your customers tend to leverage those methods, and therefore might like to know how they perform as well? Open to ideas, but we can start smaller with just the GNSS-specific additional data. It looks like we only have gnss-specific details in the location_data_details struct, so maybe expanding that to have wifi and cellular info would be a worthwhile but separate effort.

Reply:
Yes, customers also use especially cellular positioning because HW always supports that together with GNSS. Wi-Fi is getting more and more attraction but requires additional HW.